readme.md
# Staff Picks
[![GitHub version](https://badge.fury.io/gh/nypl%2Fstaff-picks.svg)](https://badge.fury.io/gh/nypl%2Fstaff-picks)
[![Build Status](https://travis-ci.org/NYPL/staff-picks.svg?branch=master)](https://travis-ci.org/NYPL/staff-picks)
[![dependencies Status](https://david-dm.org/nypl/staff-picks/status.svg)](https://david-dm.org/nypl/staff-picks)
[![devDependencies Status](https://david-dm.org/nypl/staff-picks/dev-status.svg)](https://david-dm.org/nypl/staff-picks?type=dev)

## Technology
* Universal React
* [Alt](http://alt.js.org/)/Iso as the Flux implementation
* Webpack & Webpack Dev Server
* ES6 and Babel
* ESLint with [Airbnb's config](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb)
* Unit Testing with [Mocha](https://mochajs.org/) and [Enzyme](http://airbnb.io/enzyme/)
* Express Server

## Requirements
`clientId` and `clientSecret` environment variables are **required** and available on AWS via Parameter Store

## Installation
Install all npm dependencies listed under `package.json`
```sh
$ npm install
```

## Running the Application
Before running the application, the required variables are indicated in .env.sample at the application root. Please contact the NYPL Digital Department for any help with these. Copy .env.sample to .env and overwrite the placeholder values with the correct environment variables. You are now ready to run the application.

### Development Mode
We use Webpack to fire off a hot-reloading development server. This allows for continuous code changes without the need to refresh your browser.

You can run the application in development mode by running this on the command line:

```sh
npm run dev-api-start
```

This will use the environment variables you have set in `.env`. Or, if you need to use different credentials, indicate these prior to the npm run command which will override the settings in .env:

```sh
$ CLIENT_ID=[clientId] CLIENT_SECRET=[clientSecret] npm run dev-api-start
```

> Navigate to http://localhost:3001/books-music-dvds/recommendations/best-books/ya or http://localhost:3001/books-music-dvds/recommendations/best-books/childrens

### Production Mode
To run locally in production mode using the `development` API, run:

```sh
$ NODE_ENV=production npm run dev-api-start
```

To use the `production` API, run:
```sh
$ NODE_ENV=production npm run prod-api-start
```

Notice that just as running in development mode, you can still override the credential settings in `.env` by adding `CLIENT_ID=[clientId] CLIENT_SECRET=[clientSecret]` to your production mode command line.

## GIT Workflow
We follow a [feature-branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) workflow. If you need to introduce/update the application code, you `SHOULD`:

* Create a new branch off the `development` branch
* Send a pull request pointing to the `development` branch upon completion
* Once the pull request is approved, it should be merged into the `development` branch
* Travis CI is setup to automatically build and deploy the `development` branch on our Elastic Beanstalk Development server
* If there are several pull requests in process, a release should be scheduled by merging all completed pull requests into the `development` branch
* When a release is scheduled to be deployed, the `development` branch `SHOULD` be merged into the `qa` branch
* Travis CI will build and deploy the `qa` branch into our Elastic Beanstalk QA server
* Upon a successful verification of the application on our staging (QA) environment, the `qa` branch `SHOULD` be merged into the `master` branch
* Travis CI will build and deploy the `master` branch into our Elastic Beanstalk Production server
* All releases merged into `master` `MUST` be tagged and pushed to Github with their corresponding `version`

## Testing
### Unit Tests
Unit tests are currently written in the [Mocha](https://mochajs.org/) testing framework. [Chai](http://chaijs.com/) is used as the assertion library and [Enzyme](http://airbnb.io/enzyme/) is used as the React testing utility.

The tests can be found in the `./test` folder.

To run all the tests once, run
```sh
$ npm run test
```

To run the tests continuously for active development, run
```sh
$ npm run test-dev
```

To run a specific test file, run
```sh
$ npm run test-file test/unit/TabElement.test.js
```

### Code Coverage
[Istanbul](https://istanbul.js.org/) is used for checking code coverage.

To run the code coverage tool and view a quick report, run

    $ npm run coverage

To run the code coverage tool and view a better report, run

    $ npm run coverage-report

This last command will create a folder called `coverage` in the root directory. You can open up `coverage/lcov-report/index.html` in a browser to see more details about what lines of codes have not been tested.

## AWS Elastic Beanstalk Configuration
All of the AWS Elastic Beanstalk configuration files exist in the `.ebextensions` directory. AWS will run all config files in numerical order based on the prefix `{NUM}_{FILENAME}.config` naming convention.

You must have AWS CLI installed on your machine with the proper AWS Profiles.

> Note: Please use the instance profile of `_cloudwatchable-beanstalk_`. It is configured with all of the permissions necessary for a traditional or Docker-flavored Beanstalk machine that enables logging to CloudWatch.

### AWS Application Settings
By using the `aws-cli`, developers can deploy the application to the AWS application environments listed below:

| AWS Profile | Application Name | Environment |
|---|---|---|
| `nypl-digital-dev` | `nypl-staff-picks-app` | **QA**: `staff-picks-qa` <br><br> **Production**: `staff-picks-production` |
| `nypl-sandbox` | `nypl-staff-picks-app` | **Development**: `staff-picks-development`

> Note: All QA and Development servers should be configured with only 1 instance. Production servers are typically setup with auto-scaling supporting 2 or more instances.

### Initialization

To initialize the application on AWS via the `aws-cli` run:

```bash
 eb init -i --profile ${your AWS profile}
```
### Creation of Environment

Initially we want to create the app on AWS Elastic Beanstalk by using this command:
```bash
eb create ${environment name} \
  --instance_type ${size of instance} (Ex: t2.small) \
  --instance_profile cloudwatchable-beanstalk \
  --scale ${amount of instances} (Ex: 1) \
  --cname ${cname prefix} (Ex: XXX.us-east-1.elasticbeanstalk.com) \
  --vpc.id ${ask for custom vpc_id} \
  --vpc.ec2subnets ${privateSubnetId1,privateSubnetId2} \
  --vpc.elbsubnets ${publicSubnetId1,publicSubnetId2} \
  --vpc.elbpublic (exposes the IP to the public) \
  --keyname ${ssh keyname} (allows the given ssh keyname to have ssh access) \
  --profile ${your AWS profile}
```
> Note: This step is only required upon creation and should executed once.

### Deployment
#### AWS Deployment with Continuous Integration & Delivery
Travis CI is configured to run our build and deployment process on AWS.

Our Travis CI/CD pipeline will execute the following steps for each deployment trigger:
* Run unit test coverage
* Run the npm task to build the distribution assets
* Execute the `deploy` hook only for `development`, `qa` and `master` branches to adhere to our AWS Elastic Beanstalk `development`, `qa` and `production` servers
* Developers _do not_ need to manually deploy the application unless an error occurred via Travis

#### Manual/Custom Deployment
Once the application has been created via `eb create ...` all subsequent updates and deployments should use the following command:

```bash
eb deploy ${environment name} --profile ${your AWS profile}
```

#### Switching Loadbalancer settings between accounts

Loadbalancer settings are required for HTTPS to run behind the load balancers for Elastic Beanstalk. Each Loadbalancer setting is unique to each AWS profile, so switching configuration files is required when deploying to different AWS profiles.

| AWS Profile | Configuration File | 
|---|---|
| `nypl-digital-dev` | `https-nypl-digital-dev.config` | 
| `nypl-sandbox` | `https-nypl-sandbox.config` | 

When `development` is ready to merge into `qa`, only include `https-nypl-digital-dev.config` within the `.ebextensions` directory. Since CI/CD is being performed via Travis CI, run the following commands on the local qa branch after a merge:

```bash
git checkout qa
git merge development
git rm .ebextensions/https-nypl-sandbox.config
git add .ebextensions/https-nypl-digital-dev.config
git commit
git push origin qa
```

At `git push origin qa`, Travis CI will deploy to the QA server with the correct credentials.

This procedure is not needed for deployment to `production` environment when merging from `qa`. The `qa` branch uses the same credentials from `nypl-digital-dev` account, and will be reused by `production` environment.

For details, please see [NYPL common instructions on Elastic Beanstalk](https://github.com/NYPL/aws/blob/master/common/elasticbeanstalk.md).

### KMS Environment Variables
Staff Picks and the API where the data is fetched from are currently deployed on NYPl's AWS instance. In order to fetch data, we are using the `@nypl/nypl-data-api-client` to make requests to the API with an authentication token. You can find the [full documentation here](https://www.npmjs.com/package/@nypl/nypl-data-api-client), but to be brief, we need a client id, a client secret, and a token url to authenticate.

We currently have a process in NYPL Digital to get said pair of authentication strings. To be more secure in our AWS environment, we are encrypting those variables with AWS KMS. To encrypt the strings, make sure you have the `aws-cli` tool available on your computer and then run:

```
  $ aws kms encrypt --key-id [your-aws-key-id] --plaintext [string-to-encrypt] --output text --query CiphertextBlob --profile [name-of-your-aws-profile]
```

Running this command will output an base64 encrypted string which will be used as the `KMS_ENV` environment variable in AWS Elastic Beanstalk. The app will read the variable, and use the `aws-sdk` library to decrypt it. Once we decrypt it, we can use those values with the `@nypl/nypl-data-api-client` to connect to the API.

As stated above, you will need the client id and secret when running the app locally, but they do not need to be encrypted. The encrypted values will be use in the AWS instances as environment variables and the app will pick those up and use them since the `KMS_ENV` variable will be set to `encrypted` in the Beanstalk instances.

If you would like to use the encrypted variables, you can start the app using:

```
  $ NODE_ENV=production clientId=[encryptedClientId] clientSecret=[encryptedClientSecret] npm run prod-api-start
```


CHANGELOG.md
## CHANGELOG

### 4.0.2
> Update the header component to 2.4.10.

### 4.0.1
> Updating @nypl/dgx-react-footer version to 0.5.0 and @nypl/dgx-header-component to 2.4.8.
> Updating 404 page content to include a link back to /staff-picks.
> Updating additional info link to include 'formats for'.
> Adds npm pacakge pluralize v7.0.0 to handle singular and plural word forms for titles and book counts.

### 4.0.0
> Update the routes. The base route is now `/books-music-movies/recommendations/best-books/`. And `staff-picks`, `ya`, `childrens` are three lists on the same level of the structure.
> Update the routes for supporting the old URLs for individual books. Now the links no longer lead to a modal but scroll down and focus on the particular book in the list.
> Add an `.env.example` and update the related README.md so now the developer can set a `.env` to run the application locally.
> Add a list selector component that consists of two dropdown menus for choosing both seasons and audience from Staff Picks, or only seasons from Best YA Books and Best Children Books.
> Update hero elements and styles based on different list types. Staff Picks and Children/YA lists now have different heros.
> Add lazy load when loading images for better performance.
> Add Winston and logging.

### 3.2.6
> Updating Travis CI configuration for CD to AWS Elastic Beanstalk

### 3.2.5
> Update the header component to 2.4.7.
> Added OptinMonster for advocacy 2018.

### 3.2.4
> Added Travis CI configuration for AWS EB
> Integrated swap-space configuration in ./ebextensions
> Updated README to include AWS CI/CD workflow

### 3.2.3
> Update the header component to 2.4.5.

### 3.2.2
> Update the header component to 2.4.2.

### 3.2.1
> Update the header to v2.4.1 and footer to v0.4.1.

### 3.2.0
> Update to the header, v2.4.0.

### 3.1.0
> Updating dgx-react-ga related functions for Header Component and dgx-react-ga updates.

### 3.0.3
> Updating URL of hero images to be relative.

### 3.0.2
> Updated header to v2.3.0 -- Includes FundraisingBanner integration

### 3.0.1
> Adding Twitter and FB OG tags for images.
> Small style update to the # books found in the sidebar.

### 3.0.0
> 2017 Best Books for Kids and Best Books for Teens!
> We've refactored the base Node/React application based off [Discovery](https://github.com/NYPL-discovery/discovery-front-end), added unit tests, cleaned up and updated npm packages, completed accessibility and cross-browser reviews, and updated the routes for the next big version of Staff Picks. This initial re-implementation kicks off with the 2017 annual lists for kids and teens and will later incorporate the rest of the data. The data is now coming from NYPL Digital's API platform located on AWS. The app is also updated to use the [Design Toolkit](https://github.com/NYPL/design-toolkit)! Making huge progress to unify our designs and conventions.

### 2.5.0
> Removing unnecessary npm packages, removing the modal, and updating routes. Also an update to React Router.

### 2.4.0
> Adding unit tests, updating a few npm packages, and updating readme.

### v2.3.1
> Updating the Header to version 2.1.1.

### v2.3.0
> Adding Feature Flag for displaying the List view by default, through Optimizely.

### v2.2.2
> Updating the Header to version 2.1.0.

### v2.2.1
> Fix to time/season display.

### v2.2.0
> Updated to React 15.

### v2.1.11
> Updated the logic in BookFilters.jsx to render one list of tags starting from summer 2016.

### v2.1.10
> Updated the Header Component to v1.5.5. The updates include integrating the log in related functions with login server, removing console loggings for patron token expiration, and turning off the feature flag of OAuth Login and set it as default.

### v2.1.9
> Updated the variable check for "id" in ApiRoutes.js to fix the bug that the routes can't deal with dates.

> Updated the variable check for "catalogSlug" in BookContent.jsx.

### v2.1.8
> Updated the Header Component to v1.5.1. The update includes HTTPS fix and the JavaScript fallback for the log in button on the Header Component.

### v2.1.7
#### Added
> Enabled Feature Flags plugin on the client-side and added Optimizely script in the index.ejs file.


appConfig.js
const config = {
  port: 3001,
  webpackDevServerPort: 3000,
  baseUrl: '/books-music-movies/recommendations/best-books',
  base404: '/books-music-movies/recommendations/best-books/404',
  baseApiUrl: '/books-music-movies/recommendations/best-books/api/',
  metaTags: {
    ya: [
      { property: 'og:type', content: 'website' },
      { property: 'og:site_name', content: 'Best Books for Teens' },
      { property: 'og:title', content: 'Best Books for Teens' },
      { property: 'og:description', content: 'Explore The New York Public Library\'s annual ' +
        'selection of outstanding young adult titles.' },
      { property: 'og:url', content: 'https://www.nypl.org/books-music-movies/recommendations' +
        '/best-books/ya' },
      { property: 'og:image', content: 'https://d140u095r09w96.cloudfront.net/sites/default/' +
        'files/teens-600x335.jpg' },
      { property: 'og:image:alt', content: 'Featured title: The Wendy Project by Melissa Jane ' +
        'Osborne, illustrated by Veronica Fish' },
      { property: 'og:image:width', content: '335' },
      { property: 'og:image:height', content: '600' },
      { name: 'twitter:title', content: 'Best Books for Teens' },
      { name: 'twitter:card', content: 'summary_large_image' },
      { name: 'twitter:site', content: '@nypl' },
      { name: 'twitter:creator', content: '@nypl' },
      { name: 'twitter:description', content: 'Explore The New York Public Library\'s annual ' +
        'selection of outstanding young adult titles.' },
      { name: 'twitter:image', content: 'https://d140u095r09w96.cloudfront.net/sites/default/' +
        'files/teens-600x335.jpg' },
      { name: 'twitter:image:alt', content: 'Featured title: The Wendy Project by Melissa ' +
        'Jane Osborne, illustrated by Veronica Fish' },
    ],
    childrens: [
      { property: 'og:type', content: 'website' },
      { property: 'og:site_name', content: 'Best Books for Kids' },
      { property: 'og:title', content: 'Best Books for Kids' },
      { property: 'og:description', content: 'Explore The New York Public Library\'s annual ' +
        'selection of outstanding children\'s titles.' },
      { property: 'og:url', content: 'https://www.nypl.org/books-music-movies/recommendations' +
        '/best-books/childrens' },
      { property: 'og:image', content: 'https://d140u095r09w96.cloudfront.net/sites/default/' +
        'files/kids-600x335.jpg' },
      { property: 'og:image:alt', content: 'Featured title: The Book of Mistakes by ' +
        'Corinna Luyken' },
      { property: 'og:image:width', content: '335' },
      { property: 'og:image:height', content: '600' },
      { name: 'twitter:title', content: 'Best Books for Kids' },
      { name: 'twitter:card', content: 'summary_large_image' },
      { name: 'twitter:site', content: '@nypl' },
      { name: 'twitter:creator', content: '@nypl' },
      { name: 'twitter:description', content: 'Explore The New York Public Library\'s annual ' +
        'selection of outstanding children\'s titles.' },
      { name: 'twitter:image', content: 'https://d140u095r09w96.cloudfront.net/sites/default/' +
        'files/kids-600x335.jpg' },
      { name: 'twitter:image:alt', content: 'Featured title: The Book of Mistakes by ' +
        'Corinna Luyken' },
    ],
    'staff-picks': [
      { property: 'og:type', content: 'website' },
      { property: 'og:site_name', content: 'Staff Picks' },
      { property: 'og:title', content: 'Staff Picks' },
      { property: 'og:description', content: 'Nobody loves books more than our experts. Browse and filter hundreds of their favorites.' },
      { property: 'og:url', content: 'https://www.nypl.org/books-music-movies/recommendations' +
        '/best-books/staff-picks' },
      { property: 'og:image', content: 'https://www.nypl.org/books-music-movies/recommendations' +
        '/best-books/src/client/images/shelftalker.4.2.png' },
      { property: 'og:image:alt', content: 'Featured title: Anathem by Neal Stephenson' },
      { property: 'og:image:width', content: '335' },
      { property: 'og:image:height', content: '600' },
      { name: 'twitter:title', content: 'Staff Picks' },
      { name: 'twitter:card', content: 'summary_large_image' },
      { name: 'twitter:site', content: '@nypl' },
      { name: 'twitter:creator', content: '@nypl' },
      { name: 'twitter:description', content: 'Nobody loves books more than our experts. Browse and filter hundreds of their favorites.' },
      { name: 'twitter:image', content: 'https://www.nypl.org/books-music-movies/recommendations' +
        '/best-books/src/client/images/shelftalker.4.2.png' },
      { name: 'twitter:image:alt', content: 'Featured title: Anathem by Neal Stephenson' },
    ],
  },
  seasons: {
    Spring: ['March', 'April', 'May'],
    Summer: ['June', 'July', 'August'],
    Fall: ['September', 'October', 'November'],
    Winter: ['December', 'January', 'February'],
  },
  recommendationsLink: {
    url: 'https://www.nypl.org/books-music-movies/recommendations',
    label: 'Recommendations',
  },
  aboutUrls: {
    print: 'https://www.nypl.org/accessibility/print-disabilities',
    about: 'https://www.nypl.org/books-music-movies/recommendations/about/annual-lists',
  },
  publicKey:
    '-----BEGIN PUBLIC KEY-----\n' +
    'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA44ilHg/PxcJYsISHMRyo\n' +
    'xsmez178qZpkJVXg7rOMVTLZuf05an7Pl+lX4nw/rqcvGQDXyrimciLgLkWu00xh\n' +
    'm6h6klTeJSNq2DgseF8OMw2olfuBKq1NBQ/vC8U0l5NJu34oSN4/iipgpovqAHHB\n' +
    'GV4zDt0EWSXE5xpnBWi+w1NMAX/muB2QRfRxkkhueDkAmwKvz5MXJPay7FB/WRjf\n' +
    '+7r2EN78x5iQKyCw0tpEZ5hpBX831SEnVULCnpFOcJWMPLdg0Ff6tBmgDxKQBVFI\n' +
    'Q9RrzMLTqxKnVVn2+hVpk4F/8tMsGCdd4s/AJqEQBy5lsq7ji1B63XYqi5fc1SnJ\n' +
    'EQIDAQAB\n' +
    '-----END PUBLIC KEY-----',
  heroData: {
    staffPicks: {
      category: '',
      header: 'Staff Picks',
      description: 'Nobody loves books more than our experts. ' +
        'Browse and filter hundreds of their favorites.',
      heroImageUrl: '',
      sectionTitle: 'BOOKS/MUSIC/MOVIES',
    },
    ya: {
      category: '',
      header: 'Best Books for Teens',
      description: 'Explore our annual selection of outstanding young adult titles.',
      heroImageUrl: 'src/client/images/desktop.teens.cover.2017.png',
      sectionTitle: '',
    },
    childrens: {
      category: '',
      header: 'Best Books for Kids',
      description: 'Explore our annual selection of outstanding children\'s titles.',
      heroImageUrl: 'src/client/images/desktop.kids.cover.2017.png',
      sectionTitle: '',
    },
  },
  pageTitle: {
    ya: 'Best Books for Teens | NYPL',
    childrens: 'Best Books for Kids | NYPL',
    'staff-picks': 'NYPL | Staff Picks',
  },
  requestUrlsText: {
    catalog: 'Request Book',
    ebook: 'Request E-Book',
  },
  socialSharing: {
    facebook: 'https://www.facebook.com/sharer/sharer.php?u=',
    twitter: 'https://twitter.com/intent/tweet?',
  },
  staffPicksListOptions: {
    season: {
      fieldsetName: 'season',
      currentValue: '',
      options: [],
    },
    audience: {
      fieldsetName: 'audience',
      currentValue: '',
      options: [
        { name: 'Adult', value: 'Adult' },
        { name: 'Young Adult', value: 'YA' },
        { name: 'Children', value: 'Children' },
      ],
    },
  },
  annualListOptions: {
    season: {
      fieldsetName: 'year',
      currentValue: '',
      options: [],
    },
  },
  audienceMap: {
    Adult: 'Adult',
    YA: 'Young Adult',
    Children: 'Children',
  },
  niceLabelMap: {
    'staff-picks': 'Staff Picks',
    ya: 'YA',
    teens: 'YA',
    kids: 'Childrens',
    childrens: 'Childrens',
  },
};

export default config;
